### ISSUE TITLE
Possible GUI Option

### ISSUE THREAD

**Issue by ChristianSilvermoon** [enigma9o7/smlinux#17]
I noticed when messing with the script that there's a message talking about a graphical way to edit the config file, etc.

You might be able to use `yad` for this

**Comment by enigma9o7** [comment]
Thanks for the idea.  I'm already using whiptail a few places for user interaction (including the intro menu if user does not specify a command) and zenity for a couple popups, but what I imaged is like a panel type interface, with checkboxes or sliders for toggle settings and drop-down or radiolists etc for selections, etc, oldschool ugly but functional UI type thing.  Of course doesnt have to be ugly, but with zenity/whiptail/dialog and I think yad they're more for single selections.  I dont want a million steps, but having a panel with all the options would be better than opening a text editor like I do now.  Plus it could integrate the command menu smlinux has nowt, basically one simple GUI but exact same script/fucntion, but easier to use.

**Comment by PricklySaguaro** [comment]
I'm currently learning Qt 5 for one of my classes (or, to be specific, [Python bindings for Qt](https://pypi.org/project/PyQt5/)). I'm not a designer by any means, but making an ugly yet functional GUI for configuration would be trivial for me. I can write one if you'd like.

**Comment by enigma9o7** [comment]
That's exactly the kind of thing I imagined, python qt free cross platform, I haven't learned them myself but could use/adapt a working example if you can do what you describe.  Right now the smlinux script calls a text editor to edit its config file (which is formatted as shell script).   The easiest way to quickly integrate a GUI into existing smlinux would be to replace the text editor with a GUI designed to edit the config file.  From that point I believe I could figure out how to further integrate and improve, but that much would be a great contribution if you (or anyone) could do that.

**Comment by PricklySaguaro** [comment]
> I haven't learned them myself but could use/adapt a working example if you can do what you describe.

Yeah, you're right on the money there. Qt (especially with PyQt) is pretty easy, so you'd have no problems adding stuff once I get the initial GUI working.

> The easiest way to quickly integrate a GUI into existing smlinux would be to replace the text editor with a GUI designed to edit the config file.

That was my plan. That being said, when I do my PR, would you like me to try and integrate it, or do you want it to just be the GUI Python script so you can integrate it how you like?

I estimate it'd take less than a day for me to write it, and 99% of the work would be parsing the existing config file. Since my schedule is a little funky right now, I should have something ready before the end of the week.

**Comment by enigma9o7** [comment]
Thanks!
Implementing it into the script itself I expect to be fairly trivial, so whichever you prefer.   Assuming it works, I see no reason to not start using it right away.... unless it requires a dependency that wasn't required before etc, in which case I'll deal with that...

**Comment by PricklySaguaro** [comment]
> unless it requires a dependency that wasn't required before

Now that you mention it, it will require a dependency that wasn't required before. So I'll just make a PR with the Python script so you can update the dependency list before integrating it.

**Comment by PricklySaguaro** [comment]
Okay, I made a PR. I wrote it in such a way that you don't really need knowledge of Qt5 to add options, so it should be easy to modify for any future versions of smlinux. Also, you were right: implementing it into the script was easy, so I went ahead and did that. It handles the dependency of PyQt5 using pip3. You had some code in the script that does the same thing for numpy, so I copied that.

**Comment by enigma9o7** [comment]
Gotta say thanks again!  I've managed to figure out the basics of rearranging the GUI, and how to deal with adding and removing fields, and have tested in a few distro/VM and got the dependencies sorted out.  I understand it needs PyQt5.  Does it also need numpy, I had assumed yes from the beginning cuz of the way you'd made it install numpy first, but now I'm not sure.  Before, the only time the script installed/used numpy was if the user provided a rom in another format besides z64, because make only accepts z64 so smlinux uses somebody elses python script to convert i in those cases....

I am hoping you can point me in the direction of how I can learn to modify it further.   For example, three changes I'd like to make.

1) Add a folder select dialog (for the base folder).

2) Compress the UI.  Right now it comes up bigger than 1000x700 which is too big for small resolutions (like 1024x768 1280x720 1368x768 etc) plus it just looks too spaced out.   By staring at the code I don't see anything about how it is layed out or how to control the spacing between fields.

3) Accept enter for save/continue, right now requires mouse only as far as I can tell.

I also saw what you mean about older python doesn't arrange them the same as newer python, wwhen I tested in ubuntu 16 32-bit VM with python 3.5).     I'm guessing nothing we can do about that (besides update python)...

I am unfamiliar with QT Creator but I know it's a tool for this kinda thing.  Is there a way to use it with this?

**Comment by PricklySaguaro** [comment]
Hello again!

As far as I'm aware, it does _not_ require numpy. I had copied and modified the code for installing numpy from the other section of the script in my original attempt at integrating the GUI. I thought I had modified it to install PyQt5 instead. Guess not?

All 3 things you request are possible.
1. With the way I have the configuration GUI written, adding a folder select dialog would require implementing another configuration type. It would use the `QFileDialog.getOpenFileName()` method.
2. You can change the default window size using the `self.setGeometry()` method in `ConfigWindow`'s init function. I have adjusted spacing of widgets before, but I don't recall how to do it off the top of my head. I know QWidgets have a default spacing and parent widgets inherit this. Something to do with a "size hint?"
3. The `QShortcut` class might work here.

I have no experience with Qt Creator so I can't comment on it.

I could implement this myself, but I'm busy with school for a couple weeks as the semester comes to a close. If you're willing to wait a bit, I'd be more than happy to implement all of this.
